types/bluebird/bluebird-tests.ts

```diff
@@ -313,29 +313,29 @@ fooOrBarProm = fooProm.caught((error: any) => {
 // - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
 
 // $ExpectType Bluebird<void | Foo>
-fooProm.catch(Error, (reason: any) => {
+fooProm.catch(Error as { new(message?: string): Error }, (reason: any) => {
```

Comment:
I don't quite understand the changes to this file. Can you explain to me why they're necessary?
(I haven't been following the latest changes in TS 3.2. 🤐)

I'm asking because because these particular tests assert the proper functionality of [one of Bluebird's features](http://bluebirdjs.com/docs/api/catch.html#filtered-catch).
As such it's important that the Bluebird typings work _even if you keep this test file as it is_.

Would it be possible to change the actual typings instead?

Reply:
`Error` actually appears to match *two* of the possible filter types: `class ErrorClass` and `function (anyOtherError)`, but the second match is incorrect. The typescript representation is

```ts
type CatchFilter<E> = (new (...args: any[]) => E) | ((error: E) => boolean)
interface ErrorConstructor {
    new(message?: string): Error;
    (message?: string): Error;
}
```

Inferring from the construct signature gives `E=Error`, and inferring from the call signature gives `E=string | undefined`. I'm not sure why pre-3.2 Typescript used the first inference instead of the second.

Reply:
The change happened https://github.com/Microsoft/TypeScript/pull/27028, which changed the way co- and contra-variant inferences work so that contravariant inferences (callback parameters like `(error: E) => boolean`) are preferred over covariant inferences (return positions like `new (...args:any[]) => E`).

I think this is the right change for most cases, so I'll see if I can find a workaround for this case.

Reply:
Ah I see... That's of course very unfortunate since we were relying on the `CatchFilter` preferring covariants / the constructor signature. 😟
It'd be nice if we could come up with a solution that properly resolves this. 🤔
I'll also try to come up with something as well, although I'm not as versed in typescript as I used to.

Reply:
Well, I can manually promote the constructor inferences by breaking `catch` into *more* overloads. I'm trying that now to make sure it works and see how ugly it is.

Reply:
If it works I'm absolutely happy with it. 🙂

Meanwhile I tried:
```ts
type CatchFilter<E> = E extends InstanceType<infer R> ? R : (((error: E) => boolean) | (object & E));
```

But it doesn't work: It still resolves to / picks the latter half and thus fails the type check.
I thought this ternary-like operator would actually behave like a classic ternary operator and be evaluated eagerly. But instead it seems that it's resolved just like the regular union operator?